storage/local/frankenstein.go

```diff
@@ -0,0 +1,183 @@
+	}
+	defer func() {
+		i.fpLocker.Unlock(fp)
+	}()
+
+	if sample.Timestamp == series.lastTime {
+		// Don't report "no-op appends", i.e. where timestamp and sample
+		// value are the same as for the last append, as they are a
+		// common occurrence when using client-side timestamps
+		// (e.g. Pushgateway or federation).
+		if sample.Timestamp == series.lastTime &&
+			series.lastSampleValueSet &&
+			sample.Value.Equal(series.lastSampleValue) {
+			return nil
+		}
+		return ErrDuplicateSampleForTimestamp // Caused by the caller.
+	}
+	if sample.Timestamp < series.lastTime {
+		return ErrOutOfOrderSample // Caused by the caller.
+	}
+	_, err = series.add(model.SamplePair{
+		Value:     sample.Value,
+		Timestamp: sample.Timestamp,
+	})
+	return err
+}
+
+func (i *Ingestor) getOrCreateSeries(metric model.Metric) (model.Fingerprint, *memorySeries, error) {
+	rawFP := metric.FastFingerprint()
+	i.fpLocker.Lock(rawFP)
+	fp := i.mapper.mapFP(rawFP, metric)
+	if fp != rawFP {
+		i.fpLocker.Unlock(rawFP)
+		i.fpLocker.Lock(fp)
+	}
+
+	series, ok := i.fpToSeries.get(fp)
+	if ok {
+		return fp, series, nil
+	}
+
+	var err error
+	series, err = newMemorySeries(metric, nil, time.Time{})
+	if err != nil {
+		// err should always be nil when chunkDescs are nil
+		panic(err)
+	}
+	i.fpToSeries.put(fp, series)
+	return fp, series, nil
+}
+
+func (i *Ingestor) Query(from, to model.Time, matchers ...*metric.LabelMatcher) (model.Matrix, error) {
+	return model.Matrix{}, nil
+}
+
+func (i *Ingestor) Stop() {
+	close(i.quit)
+	<-i.done
+}
+
+func (i *Ingestor) loop() {
+	defer close(i.done)
+	tick := time.Tick(flushCheckPeriod)
+	for {
+		select {
+		case <-tick:
+			for pair := range i.fpToSeries.iter() {
+				if err := i.flushSeries(pair.fp, pair.series); err != nil {
+					log.Errorf("Failed to flush chunks for series: %v", err)
+				}
+			}
+		case <-i.quit:
+			return
+		}
+	}
+}
+
+func (i *Ingestor) flushSeries(fp model.Fingerprint, series *memorySeries) error {
+	i.fpLocker.Lock(fp)
+
+	// Decide what chunks to flush
+	chunks := series.chunkDescs
+	if !series.headChunkClosed {
+		chunks = chunks[:len(chunks)-1]
+	}
+	if len(chunks) == 0 {
+		i.fpLocker.Unlock(fp)
+		return nil
+	}
+
+	// flush the chunks without locking the series
+	i.fpLocker.Unlock(fp)
+	if err := i.flushChunks(fp, series.metric, chunks); err != nil {
+		return err
+	}
+
+	// now remove the chunks
+	// TODO deal with iterator reading from them
+	i.fpLocker.Lock(fp)
+	series.chunkDescs = series.chunkDescs[len(chunks):]
+	i.fpLocker.Unlock(fp)
+	return nil
+}
+
+func (i *Ingestor) flushChunks(fp model.Fingerprint, metric model.Metric, chunks []*chunkDesc) error {
+	wireChunks := make([]wire.Chunk, len(chunks))
+	for _, chunk := range chunks {
+		buf := make([]byte, chunkLen)
+		if err := chunk.c.marshalToBuf(buf); err != nil {
+			return err
+		}
+
+		wireChunks = append(wireChunks, wire.Chunk{
+			ID:      fmt.Sprintf("%d:%d:%d", fp, chunk.chunkFirstTime, chunk.chunkLastTime),
```

Comment:
There'll be many chunks with the same contents (think of multiple constant-zero series from the same scrape, so they'll have the same timestamps too). If we want to go the content-addressable route, that could actually be ok. Then it just means that a chunk can belong to multiple time series, which will imply a different cleanup logic when chunks get deleted, etc.

Otherwise, do a combination of series fingerprint and the chunk content hash I'd say. Or any GUID scheme.